sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala

```diff
@@ -81,10 +86,220 @@ object ReorderJoin extends Rule[LogicalPlan] with PredicateHelper {
     }
   }
 
+  /**
+   * Helper function that computes the size of a table access after applying
+   * the predicates. Currently, the function returns the table size.
+   * When plan cardinality and predicate selectivity are implemented in Catalyst,
+   * the function will be refined based on these estimates.
+   */
+  private def computeBaseTableSize(
+      input: LogicalPlan,
+      conditions: Seq[Expression]): Option[BigInt] = {
+    input match {
+      case BaseTableAccess(t, cond) if t.statistics.sizeInBytes >= 0 =>
+        // To be replaced by the table cardinality, when available.
+        val tableSize = t.statistics.sizeInBytes
+
+        // Collect predicate selectivity, when available.
+        // val predicates = conditions.filter(canEvaluate(_, p)) ++ cond
+        // Compute the output cardinality = tableSize * predicates' selectivity.
+        Option(tableSize)
+
+      case _ => None
+    }
+  }
+
+  /**
+   * Helper case class to hold (plan, size) pairs.
+   */
+  private case class TableSize(plan: (LogicalPlan, InnerLike), size: Option[BigInt])
+
+  /**
+   * Checks if a star join is a selective join. A star join is assumed
+   * to be selective if (1) there are local predicates on the dimension
+   * tables and (2) the join predicates are equi joins.
+   */
+  private def isSelectiveStarJoin(
+      starJoinPlan: Seq[LogicalPlan],
+      conditions: Seq[Expression]): Boolean = {
```

Comment:
@gatorsmile Thank you for reviewing the changes. I agree with your suggestions. It’s more clear if we pass fact + dimension tables.